### PR TITLE
Support for "grafted" files

### PR DESCRIPTION
**Pull request by bbockelm**
This PR adds support for 'grafted' files - files that are added to the repository without being physically present on the scratch filesystem.

For a given file $FOO in a directory, we denote a grafted file by the presence of .cvmfsgraft-$FOO in the same directory.  The .cvmfsgraft- file is line-oriented with key-value pairs.  Currently, two keys are recognized:

size=<integer size>
checksum=<checksum value>

The checksum must be a valid CVMFS checksum, following CVMFS checksumming conventions (such as having suffixes and being applied to the compressed contents).

If grafting is used, it is the responsibility of the repository maintainer to stage the files into the repository.

See CVM-908 for more information.

### CODE REVIEW

**Comment by bbockelm** [comment]
@jblomer - can I get you to review this?

File: cvmfs/sync_item.h [hunk 1]
```diff
@@ -153,8 +162,11 @@ class SyncItem {
   mutable EntryStat scratch_stat_;
 
   bool whiteout_;
+  bool valid_graft_;
+  bool graft_marker_present_;
   std::string relative_parent_path_;
   std::string filename_;
+  ssize_t size_;
```
Comment (jblomer) [inline cvmfs/sync_item.h]: perhaps better `grafted_size_` to clarify this is only used in the context of grafted files.

File: cvmfs/sync_item.cc [hunk 2]
```diff
@@ -185,4 +192,81 @@ std::string SyncItem::GetScratchPath() const {
...
+  if (fp == NULL) {
+    if (errno != ENOENT) {
+      LogCvmfs(kLogFsTraversal, kLogWarning, "Unable to open graft file (%s): %s (errno=%d)",
+               graftfile.c_str(), strerror(errno), errno);
+    }
+    return;
+  }
+  graft_marker_present_ = true;
+  size_t len = 0;
+  ssize_t read;
```
Comment (jblomer) [inline cvmfs/sync_item.cc]: We have a few utility functions in util.cc that should be more handy here.  In particular, `GetLineFile()` and `SplitString()` should make this paragraph shorter.

File: cvmfs/sync_item.cc [hunk 3]
```diff
@@ -185,4 +192,81 @@ std::string SyncItem::GetScratchPath() const {
...
+      }
+    }
+    if (line[read-1] == '\n') {line[read-1] = '\0';}
+    char *value;
+    if (line == (value = strcasestr(line, "size="))) {
+      value += 5;
+      size_t input_len = strlen(value);
+      char *endptr=NULL;
+      errno = 0;
+      long long mysize = strtoll(value, &endptr, 10);
```
Comment (jblomer) [inline cvmfs/sync_item.cc]: There is also `String2Uint64()` in util.cc

File: test/src/597-graft/main [hunk 4]
```diff
@@ -0,0 +1,48 @@
+
```
Comment (jblomer) [inline test/src/597-graft/main]: A few more things to test would be
 - a grafted file in a sub directory / nested catalog
 - graft markers for symlinks, directory entries (should perhaps result in a warning)
 - remove a grafted file

**Comment by jblomer** [comment]
Except for minor things and a few style violations looks fine to me.  Any other comments, @reneme?

A principal issue with this approach could be that it is easy to "migrate" a grafted file to a native file, by touching it in succeeding publish operations.  How should we deal with this?

File: test/src/597-graft/main (hunk 4 above)
Comment (bbockelm) [inline test/src/597-graft/main]: What's the best way to test for a warning?

**Comment by bbockelm** [comment]
Ok, I think I covered the cases requested in the unit tests!

File: test/src/597-graft/main [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+cvmfs_test_name="Grafting"
+cvmfs_test_autofs_on_startup=false
+
+verify_grafting() {
+  testfname=$1
+  local file1=$(cat /cvmfs/$CVMFS_TEST_REPO/file1)
+  local file2=$(cat $testfname)
+  if [ "x$file1 " == "x" ]; then
```
Comment (reneme) [inline test/src/597-graft/main]: is the space behind `$file1` intentional? This should never be `true`, no?

File: test/src/597-graft/main (hunk 4 above)
Comment (reneme) [inline test/src/597-graft/main]: usually I send the output of the publishing operation into a text file and then `grep` for the warning from there...

File: cvmfs/sync_item.cc [hunk 6]
```diff
@@ -185,4 +195,80 @@ std::string SyncItem::GetScratchPath() const {
   return union_engine_->scratch_path() + relative_path;
 }
 
+
+std::string SyncItem::GetGraftPath() const {
```
Comment (reneme) [inline cvmfs/sync_item.cc]: `::GetGraftMarkerPath()` would be more precise.

**Comment by bbockelm** [comment]
@reneme - anything else?

**Comment by jblomer** [comment]
Merged with small adjustments.

**Pull request merged**